From 23222cd8a3d3bfaf851e9522780037c566f29513 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Fri, 13 Jun 2014 14:36:49 -0400
Subject: [block] use __blk_end_request_all to free bios and also call rq->end_io

Message-id: <1402670238-13916-122-git-send-email-snitzer@redhat.com>
Patchwork-id: 83876
O-Subject: [RHEL7.1 PATCH 121/150] block: use __blk_end_request_all to free bios and also call rq->end_io
Bugzilla: 1105204
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Jeff Moyer <jmoyer@redhat.com>

BZ: 1105204

Upstream commit 7652113c2f508b1c8176640dcd034730fe79bc48
Author: Mike Christie <michaelc@cs.wisc.edu>
Date:   Wed Sep 18 08:33:55 2013 -0600

    If the queue is dying then we only call the rq->end_io callout.
    This leaves bios setup on the request, because the caller assumes when
    the blk_execute_rq_nowait/blk_execute_rq call has completed that
    the rq->bios have been cleaned up.

    This patch has blk_execute_rq_nowait use __blk_end_request_all
    to free bios and also call rq->end_io.

    Signed-off-by: Mike Christie <michaelc@cs.wisc.edu>
    Signed-off-by: Jens Axboe <axboe@kernel.dk>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/block/blk-exec.c b/block/blk-exec.c
index ead24e6..dbf4502 100644
--- a/block/blk-exec.c
+++ b/block/blk-exec.c
@@ -78,9 +78,9 @@ void blk_execute_rq_nowait(struct request_queue *q, struct gendisk *bd_disk,
  spin_lock_irq(q->queue_lock);
 
  if (unlikely(blk_queue_dying(q))) {
+  rq->cmd_flags |= REQ_QUIET; 
   rq->errors = -ENXIO;
-  if (rq->end_io)
-   rq->end_io(rq, rq->errors);
+  __blk_end_request_all(rq, rq->errors);
   spin_unlock_irq(q->queue_lock);
   return;
  }
-- 
1.7.1